woocommerce/woocommerce-ios
# REST API: Update SiteCredentialLoginUseCase to handle direct login

Closes: #9143 


## Description

This PR updates how site credential login is handled to be able to show specific error messages for failure cases.

Previously, we handled authentication passively through a call to check Jetpack with cookie authentication. `WordPressOrgNetwork` attempts to make the request, and if it fails with 401 code it tries to log in. If the login fails, the request is not retried and hence fails with 401. The `SiteCredentialLoginUseCase` detects 401 and always throw the incorrect credential error.

This PR changes this login mechanism following Android's solution mentioned in pe5sF9-1iQ-p2:
- For simplicity, the login is not called through stores/action and instead handled with `URLSession` right in the use case.
- First, a login request with a redirect URL is triggered. The response is then handled for different errors:
  - Status code 404: the login URL is not accessible (maybe customized).
  - Status code 200 and there's a div for `login_error` in the login page, scrape the message and show it in the alert.
  - Otherwise, show a message for an unexpected response.
  - Other status codes: show a message for unexpected status codes. 
  - If a redirect is detected through the session delegate, cancel the request.
  - When the request is detected to be cancelled, check access to wp-admin page. The request should succeed upon status code 200; if it's 404 the page is customized and other codes mean failure too.

The missing tracking event for the failed site credential login has also been added. The underlying error code for each error is used to track the `error_code` property for the failure, which will be used later to segment the failure metrics.

## Testing instructions

Testing customized wp-login.php URL:
- Install [WP Hide](https://wordpress.org/plugins/wp-hide-security-enhancer/) plugin on your test site.
- In wp-admin side bar, navigate to WP Hide > Hide Admin.
- Enter your custom login URL and select Show in "Block default wp-login.php" option and save the changes.
- On the app, try log in to your test site with any credentials.
- Notice an alert saying "Login failed because the access to the wp-login.php page on your site is blocked."
- In Xcode console notice a log: `🔵 Tracked login_site_credentials_login_failed, properties: [AnyHashable("error_domain"): "WooCommerce.SiteCredentialLoginError", AnyHashable("step"): "authentication", AnyHashable("error_description"): "...", AnyHashable("error_code"): "404"]`

Testing customized wp-admin URL:
- Install [WP Hide](https://wordpress.org/plugins/wp-hide-security-enhancer/) plugin on your test site.
- In wp-admin side bar, navigate to WP Hide > Hide Admin > Admin URL.
- Enter your custom admin URL and select Show in "Block default Admin URL" option and save the changes.
- On the app, try log in to your test site with correct credentials.
- Notice an alert saying "Login failed because the access to the wp-admin page on your site is blocked."
- In Xcode console notice a log: `🔵 Tracked login_site_credentials_login_failed, properties: [AnyHashable("step"): "authentication", AnyHashable("error_domain"): "WooCommerce.SiteCredentialLoginError", AnyHashable("error_code"): "404", AnyHashable("error_description"): "..."]`

Testing sites with custom javascript on their wp-login.php page:
- On the app, try log in to a site with custom javascript on their login page (or try this p1678258552518089/1678174641.255659-slack-C03URUK5C).
- Enter any random credentials.
- Notice an error message: "Login failed with an unexpected response from your site. We are working on fixing this issue."
- In Xcode console: `🔵 Tracked login_site_credentials_login_failed, properties: [AnyHashable("step"): "authentication", AnyHashable("error_description"): "...", AnyHashable("error_domain"): "WooCommerce.SiteCredentialLoginError", AnyHashable("error_code"): "-1"]`

Testing incorrect redirect from wp-login.php:
- Try log in to a site with custom redirect for wp-login.php (e.g this one p4a5px-2RM-p2#comment-12916) and any credentials.
- Notice an error message: "Login failed with an unexpected response from your site. We are working on fixing this issue."
- Notice in Xcode console: `🔵 Tracked login_site_credentials_login_failed, properties: [AnyHashable("error_domain"): "WooCommerce.SiteCredentialLoginError", AnyHashable("error_code"): "-1", AnyHashable("error_description"): "...", AnyHashable("step"): "authentication"]`

Testing sites with captcha on their wp-login.php page:
- Install the [Login reCaptcha](https://wordpress.org/plugins/login-recaptcha/) plugin on your test site and configure it.
- On the app, try log in to your site with any credentials.
- Notice that login fails with an error message that matches the error text on your login page (e.g: "ERROR : Please check the ReCaptcha box")
- Notice in Xcode console: `🔵 Tracked login_site_credentials_login_failed, properties: [AnyHashable("error_code"): "401", AnyHashable("error_domain"): "WooCommerce.SiteCredentialLoginError", AnyHashable("step"): "authentication", AnyHashable("error_description"): "..."]`

Testing incorrect credentials:
- Disable all security plugins on your site.
- Log in to your site with incorrect credentials.
- Notice that login fails with an error message that matches the error text on your login page (e.g: "Error: The username sdfsdf is not registered on this site. If you are unsure of your username, try your email address instead.")
- Notice in Xcode console: `🔵 Tracked login_site_credentials_login_failed, properties: [AnyHashable("error_domain"): "WooCommerce.SiteCredentialLoginError", AnyHashable("error_code"): "401", AnyHashable("error_description"): "...", AnyHashable("step"): "authentication"]`
- Try again with the correct credentials, you should be logged in successfully.

## Screenshots

<img src="https://user-images.githubusercontent.com/5533851/225230172-24756bdf-f10a-40b4-905c-5b6728987527.png" width=320 />

---
